### PR TITLE
Compiler warning fixes

### PR DESCRIPTION
**Pull request by geertu**
This pull request fixes compiler warnings, and enables checking of printf()-style format strings

The first commit is also included in pull request #775.

### CODE REVIEW

**Comment by enjoy-digital** [comment]
Thanks @geertu, it looks fine. I'll have a closer look at it later. @gsomlo in case you have time, would you mind also reviewing this?

File: litex/soc/software/liblitesdcard/sdcard.c [hunk 1]
```diff
@@ -504,7 +504,7 @@ void sdcard_write(uint32_t block, uint32_t count, uint8_t* buf)
 	while (count--) {
 		/* Initialize DMA Reader */
 		sdmem2block_dma_enable_write(0);
-		sdmem2block_dma_base_write((uint64_t) buf);
+		sdmem2block_dma_base_write((uint64_t)(uintptr_t) buf);
```
Comment (gsomlo) [inline litex/soc/software/liblitesdcard/sdcard.c:507]: I'm building this for 64-bit Rocket, and I'm getting this warning (that I didn't use to get before):

```
 CC       sdcard.o
/home/somlo/LITEX/litex/litex/soc/software/liblitesdcard/sdcard.c: In function 'sdcard_read':
/home/somlo/LITEX/litex/litex/soc/software/liblitesdcard/sdcard.c:476:39: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  476 |  sdblock2mem_dma_base_write((uint64_t)(uintptr_t) buf);
      |                                       ^
/home/somlo/LITEX/litex/litex/soc/software/liblitesdcard/sdcard.c: In function 'sdcard_write':
/home/somlo/LITEX/litex/litex/soc/software/liblitesdcard/sdcard.c:507:40: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  507 |   sdmem2block_dma_base_write((uint64_t)(uintptr_t) buf);
      |                                        ^
```

Is there a way to silence the warnings on *both* 32- and 64-bit builds?

**Comment by gsomlo** [comment]
In addition to the sdcard/dma base addr. cast warnings, I'm also getting the following, which is new:

```
 CC       readline.o
In file included from /home/somlo/LITEX/litex/litex/soc/software/bios/readline.c:19:
/home/somlo/LITEX/litex/litex/soc/software/bios/readline.c: In function 'cread_add_char':
/home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:11: warning: field precision specifier '.*' expects argument of type 'int', but argument 2 has type 'long unsigned int' [-Wformat=]
   47 |   printf ("%.*s", n, str); \
      |           ^~~~~~
/home/somlo/LITEX/litex/litex/soc/software/bios/readline.c:173:3: note: in expansion of macro 'putnstr'
  173 |   putnstr(buf + *num, wlen);
      |   ^~~~~~~
/home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:14: note: format string is defined here
   47 |   printf ("%.*s", n, str); \
      |            ~~^~
      |              |
      |              int
/home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:11: warning: field precision specifier '.*' expects argument of type 'int', but argument 2 has type 'long unsigned int' [-Wformat=]
   47 |   printf ("%.*s", n, str); \
      |           ^~~~~~
/home/somlo/LITEX/litex/litex/soc/software/bios/readline.c:182:3: note: in expansion of macro 'putnstr'
  182 |   putnstr(buf + *num, wlen);
      |   ^~~~~~~
/home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:14: note: format string is defined here
   47 |   printf ("%.*s", n, str); \
      |            ~~^~
      |              |
      |              int
 LD       bios.elf
```
I'd have to spend a bit longer to tie them back to any specific change in this series. @geertu, hopefully this makes sense to you since you've been looking at the nitty-gritty of this more recently... :) Thanks!

File: litex/soc/software/liblitesdcard/sdcard.c (hunk 1 above)
Comment (geertu) [reply litex/soc/software/liblitesdcard/sdcard.c:507]: The typedefs for (u)intptr_t in litex/soc/software/include/base/stdint.h are wrong:

> typedef int intptr_t;
> typedef unsigned int uintptr_t;

On a 64-bit system, they should use long, not int, cfr. /usr/include/stdint.h.

**Comment by geertu** [comment]
> In addition to the sdcard/dma base addr. cast warnings, I'm also getting the following, which is new:
> 
> ```
>  CC       readline.o
> In file included from /home/somlo/LITEX/litex/litex/soc/software/bios/readline.c:19:
> /home/somlo/LITEX/litex/litex/soc/software/bios/readline.c: In function 'cread_add_char':
> /home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:11: warning: field precision specifier '.*' expects argument of type 'int', but argument 2 has type 'long unsigned int' [-Wformat=]
>    47 |   printf ("%.*s", n, str); \
>       |           ^~~~~~
> /home/somlo/LITEX/litex/litex/soc/software/bios/readline.c:173:3: note: in expansion of macro 'putnstr'
>   173 |   putnstr(buf + *num, wlen);
>       |   ^~~~~~~
> /home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:14: note: format string is defined here
>    47 |   printf ("%.*s", n, str); \
>       |            ~~^~
>       |              |
>       |              int
> /home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:11: warning: field precision specifier '.*' expects argument of type 'int', but argument 2 has type 'long unsigned int' [-Wformat=]
>    47 |   printf ("%.*s", n, str); \
>       |           ^~~~~~
> /home/somlo/LITEX/litex/litex/soc/software/bios/readline.c:182:3: note: in expansion of macro 'putnstr'
>   182 |   putnstr(buf + *num, wlen);
>       |   ^~~~~~~
> /home/somlo/LITEX/litex/litex/soc/software/bios/readline.h:47:14: note: format string is defined here
>    47 |   printf ("%.*s", n, str); \
>       |            ~~^~
>       |              |
>       |              int
>  LD       bios.elf
> ```
> 
> I'd have to spend a bit longer to tie them back to any specific change in this series. @geertu, hopefully this makes sense to you since you've been looking at the nitty-gritty of this more recently... :) Thanks!

Sorry, I missed that readline.c is not compiled on my system, as TERM_MINI is set.
At first sight, it looks like readline.c is using unsigned long for many variables that can just be unsigned int.
Will look into it later, and redraft a new PR.

Thanks for checking!

**Comment by geertu** [comment]
@enjoy-digital I have updated my branch.  This should fix all issues reported by @gsomlo 

Tested with linux-on-litex-vexriscv on OrangeCrab.
Build-tested with linux-on-litex-rocket and --with-{sdram,ethernet,etherbone,i2c,sdcard}.

Thanks!

**Review by enjoy-digital — APPROVED** [review]
Thanks @geertu, @gsomlo.

**Pull request merged**